### PR TITLE
[ntp] Clean up

### PR DESCRIPTION
**Pull request by lolodomo**
Add null annotations
Suppress one unused configutation setting (locale)
Get the default timezone from TimeZoneProvider
Map thing configuration and channel configuration to classes
Use constructor injection for the discovery service

Signed-off-by: Laurent Garnier <lg.hc@free.fr>

### CODE REVIEW

**Comment by lolodomo** [comment]
Binding fully retested with these changes.

**Comment by lolodomo** [comment]
I need to adjust the integration tests.

**Comment by lolodomo** [comment]
Integration tests now ok again.

File: bundles/org.openhab.binding.ntp/src/main/java/org/openhab/binding/ntp/internal/config/NtpThingConfiguration.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    public static final String REFRESH_NTP = "refreshNtp";
+    public static final String SERVER_PORT = "serverPort";
+    public static final String TIMEZONE = "timeZone";
+
+    private static final String DEFAULT_SERVER_HOSTNAME = "0.pool.ntp.org";
+    private static final int DEFAULT_REFRESH_INTERVAL = 60;
+    private static final int DEFAULT_REFRESH_NTP = 30;
+    private static final int DEFAULT_SERVER_PORT = 123;
+
+    public @Nullable String hostname;
```
Comment (J-N-K) [inline bundles/org.openhab.binding.ntp/src/main/java/org/openhab/binding/ntp/internal/config/NtpThingConfiguration.java:38]: Why don't you use something like 
```suggestion
    public String hostname = "0.pool.ntp.org";
```
You can skip the whole logic below then and directly access the fields in the code. If the user sets set parameter, the default is overwritten.

File: bundles/org.openhab.binding.ntp/src/main/java/org/openhab/binding/ntp/internal/config/NtpThingConfiguration.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    public @Nullable Integer refreshInterval;
```
Comment (J-N-K) [inline bundles/org.openhab.binding.ntp/src/main/java/org/openhab/binding/ntp/internal/config/NtpThingConfiguration.java:39]:
```suggestion
    public int refreshInterval = 60;
```

File: bundles/org.openhab.binding.ntp/src/main/java/org/openhab/binding/ntp/internal/discovery/NtpDiscovery.java [hunk 3]
```diff
@@ -44,12 +48,12 @@ public NtpDiscovery() throws IllegalArgumentException {
     }
 
     @Override
-    protected void activate(Map<String, Object> configProperties) {
+    protected void activate(@Nullable Map<String, @Nullable Object> configProperties) {
```
Comment (J-N-K) [inline bundles/org.openhab.binding.ntp/src/main/java/org/openhab/binding/ntp/internal/discovery/NtpDiscovery.java:51]: Why is this needed at all? It only calls the method it overrides

File: bundles/org.openhab.binding.ntp/src/main/java/org/openhab/binding/ntp/internal/handler/NtpHandler.java [hunk 4]
```diff
@@ -54,217 +54,183 @@
  * to one of the channels.
  *
  * @author Marcel Verpaalen - Initial contribution OH2 ntp binding
- * @author Thomas.Eichstaedt-Engelen OH1 ntp binding (getTime routine)
+ * @author Thomas.Eichstaedt - Engelen OH1 ntp binding (getTime routine)
```
Comment (J-N-K) [inline bundles/org.openhab.binding.ntp/src/main/java/org/openhab/binding/ntp/internal/handler/NtpHandler.java:57]: This looks wrong. Thomas's last name is "Eichstaedt-Engelen", so the spaces arounf the dash are wrong.

File: bundles/org.openhab.binding.ntp/src/main/java/org/openhab/binding/ntp/internal/handler/NtpHandler.java [hunk 5]
```diff
@@ -54,217 +54,183 @@
...
 
-    /** for logging purposes */
-    private final DateFormat SDF = new SimpleDateFormat(DATE_PATTERN_WITH_TZ);
+    private final Logger logger = LoggerFactory.getLogger(NtpHandler.class);
 
     /** for publish purposes */
     private DateTimeFormatter dateTimeFormat = DateTimeFormatter.ofPattern(DATE_PATTERN_WITH_TZ);
 
-    private final LocaleProvider localeProvider;
+    private @Nullable NtpThingConfiguration configuration;
```
Comment (J-N-K) [inline bundles/org.openhab.binding.ntp/src/main/java/org/openhab/binding/ntp/internal/handler/NtpHandler.java:76]: Since all values have reasonable defaults, you can skip null checks by using 
```suggestion
    private NtpThingConfiguration configuration = new NtpThingConfiguration();
```

File: bundles/org.openhab.binding.ntp/src/main/java/org/openhab/binding/ntp/internal/handler/NtpHandler.java [hunk 6]
```diff
@@ -54,217 +54,183 @@
...
+        logger.debug("Initializing NTP handler for '{}'.", getThing().getUID());
 
-            Configuration config = getThing().getConfiguration();
-            hostname = config.get(PROPERTY_NTP_SERVER_HOST).toString();
-            port = (BigDecimal) config.get(PROPERTY_NTP_SERVER_PORT);
-            refreshInterval = (BigDecimal) config.get(PROPERTY_REFRESH_INTERVAL);
-            refreshNtp = (BigDecimal) config.get(PROPERTY_REFRESH_NTP);
-            refreshNtpCount = 0;
+        NtpThingConfiguration config = getConfigAs(NtpThingConfiguration.class);
+        configuration = config;
```
Comment (J-N-K) [inline bundles/org.openhab.binding.ntp/src/main/java/org/openhab/binding/ntp/internal/handler/NtpHandler.java:107]: Why not use
```suggestion
        configuration = getConfigAs(NtpThingConfiguration.class);
```

File: bundles/org.openhab.binding.ntp/src/main/java/org/openhab/binding/ntp/internal/handler/NtpHandler.java [hunk 7]
```diff
@@ -54,217 +54,183 @@
...
-                            getThing().getUID(), timeZone, PROPERTY_TIMEZONE);
-                }
-            } catch (Exception e) {
-                timeZone = TimeZone.getDefault();
-                logger.debug("{} using default TZ '{}' due to an occurred exception: ", getThing().getUID(), timeZone,
-                        e);
-            }
+        refreshNtpCount = 0;
 
+        String timeZoneConfigValue = config.getTimeZone();
```
Comment (J-N-K) [inline bundles/org.openhab.binding.ntp/src/main/java/org/openhab/binding/ntp/internal/handler/NtpHandler.java:111]: If you use the defaults directly, you can access the fields here  and omit most of the null checks/function calls.

File: bundles/org.openhab.binding.ntp/src/main/java/org/openhab/binding/ntp/internal/handler/NtpHandler.java [hunk 8]
```diff
@@ -54,217 +54,183 @@
...
-                Object localeStringConfigValue = config.get(PROPERTY_LOCALE);
-                if (localeStringConfigValue != null) {
-                    locale = new Locale(localeStringConfigValue.toString());
-                } else {
-                    locale = localeProvider.getLocale();
-                    logger.debug("{} using default locale '{}', because configuration property '{}' is null.",
-                            getThing().getUID(), locale, PROPERTY_LOCALE);
-                }
+                timeZoneId = TimeZone.getTimeZone(timeZoneConfigValue).toZoneId();
             } catch (Exception e) {
```
Comment (J-N-K) [inline bundles/org.openhab.binding.ntp/src/main/java/org/openhab/binding/ntp/internal/handler/NtpHandler.java:117]: Is it necessary to catch `Exception` here?

File: bundles/org.openhab.binding.ntp/src/main/java/org/openhab/binding/ntp/internal/NtpBindingConstants.java [hunk 9]
```diff
@@ -37,14 +37,5 @@
     public static final String CHANNEL_DATE_TIME = "dateTime";
     public static final String CHANNEL_STRING = "string";
 
-    // Custom Properties
-    public static final String PROPERTY_NTP_SERVER_HOST = "hostname";
```
Comment (J-N-K) [inline bundles/org.openhab.binding.ntp/src/main/java/org/openhab/binding/ntp/internal/NtpBindingConstants.java:41]: I would prefer to keep the constants here, as that is what is most common in other bindings.

File: bundles/org.openhab.binding.ntp/src/main/java/org/openhab/binding/ntp/internal/discovery/NtpDiscovery.java (hunk 3 above)
Comment (lolodomo) [reply bundles/org.openhab.binding.ntp/src/main/java/org/openhab/binding/ntp/internal/discovery/NtpDiscovery.java:51]: I should refactor this call and rather use injection.

File: bundles/org.openhab.binding.ntp/src/main/java/org/openhab/binding/ntp/internal/config/NtpThingConfiguration.java (hunk 1 above)
Comment (lolodomo) [reply bundles/org.openhab.binding.ntp/src/main/java/org/openhab/binding/ntp/internal/config/NtpThingConfiguration.java:38]: Ok, fixed.

File: bundles/org.openhab.binding.ntp/src/main/java/org/openhab/binding/ntp/internal/config/NtpThingConfiguration.java (hunk 2 above)
Comment (lolodomo) [reply bundles/org.openhab.binding.ntp/src/main/java/org/openhab/binding/ntp/internal/config/NtpThingConfiguration.java:39]: Ok, fixed.

File: bundles/org.openhab.binding.ntp/src/main/java/org/openhab/binding/ntp/internal/handler/NtpHandler.java (hunk 4 above)
Comment (lolodomo) [reply bundles/org.openhab.binding.ntp/src/main/java/org/openhab/binding/ntp/internal/handler/NtpHandler.java:57]: There was a warning on this line but my change was stupid.

Comment (lolodomo) [reply bundles/org.openhab.binding.ntp/src/main/java/org/openhab/binding/ntp/internal/handler/NtpHandler.java:57]: Fixed

File: bundles/org.openhab.binding.ntp/src/main/java/org/openhab/binding/ntp/internal/handler/NtpHandler.java (hunk 5 above)
Comment (lolodomo) [reply bundles/org.openhab.binding.ntp/src/main/java/org/openhab/binding/ntp/internal/handler/NtpHandler.java:76]: Ok, done.

File: bundles/org.openhab.binding.ntp/src/main/java/org/openhab/binding/ntp/internal/handler/NtpHandler.java (hunk 6 above)
Comment (lolodomo) [reply bundles/org.openhab.binding.ntp/src/main/java/org/openhab/binding/ntp/internal/handler/NtpHandler.java:107]: I was to avoid null check warnings due to configuration which was nullable.
Now that configuration is non null, the local variable can be avoided.

Comment (lolodomo) [reply bundles/org.openhab.binding.ntp/src/main/java/org/openhab/binding/ntp/internal/handler/NtpHandler.java:107]: Local variable removed.

File: bundles/org.openhab.binding.ntp/src/main/java/org/openhab/binding/ntp/internal/handler/NtpHandler.java (hunk 7 above)
Comment (lolodomo) [reply bundles/org.openhab.binding.ntp/src/main/java/org/openhab/binding/ntp/internal/handler/NtpHandler.java:111]: I can't define a default for the configuration because the expected default is the timezone provided by the TimeZoneProvider.

File: bundles/org.openhab.binding.ntp/src/main/java/org/openhab/binding/ntp/internal/handler/NtpHandler.java (hunk 8 above)
Comment (lolodomo) [reply bundles/org.openhab.binding.ntp/src/main/java/org/openhab/binding/ntp/internal/handler/NtpHandler.java:117]: You're right, it is not necessary because it returns GMT if the string is invalid.
Fixed.

File: bundles/org.openhab.binding.ntp/src/main/java/org/openhab/binding/ntp/internal/NtpBindingConstants.java (hunk 9 above)
Comment (lolodomo) [reply bundles/org.openhab.binding.ntp/src/main/java/org/openhab/binding/ntp/internal/NtpBindingConstants.java:41]: I often saw the constants defined in th econfiguration class too.
But ok, I revert.

File: bundles/org.openhab.binding.ntp/src/main/java/org/openhab/binding/ntp/internal/discovery/NtpDiscovery.java (hunk 3 above)
Comment (lolodomo) [reply bundles/org.openhab.binding.ntp/src/main/java/org/openhab/binding/ntp/internal/discovery/NtpDiscovery.java]: Constructor injection done.

**Comment by lolodomo** [comment]
java.util.TimeZone is no more used.
java.time.ZoneId is rather used.

**Comment by lolodomo** [comment]
> java.util.TimeZone is no more used.
> java.time.ZoneId is rather used.

I have to test again but it is possible that parsing of the timezone configuration setting is now failing in particular cases like for example "GMT+1:30" when there is no existing zone id. In that case, I will revert to `TimeZone.getTimeZone(string)`.

**Comment by lolodomo** [comment]
And the full build is green.

**Comment by lolodomo** [comment]
This is not yet good as TimeZoneProvider is considered only at thing handler initialization.

**Comment by lolodomo** [comment]
I tested how the binding is behaving with these changes when no timezone is set on the thing side but the user changes the openHAB system timezone. And this is working well, the openHAB system timezone is correctly taken into accound when the date/time is updated by the binding.

**Comment by lolodomo** [comment]
@J-N-K : could you please finish your review or more exactly review my changes after your first review ?

**Comment by cpmeister** [comment]
@J-N-K can you take a final look?

**Comment by lolodomo** [comment]
Can we have this PR merged before 2.5.6 is released?

**Comment by cpmeister** [comment]
@J-N-K friendly ping
